From 464b7b740daacbb1959da791ea50e56156b412fe Mon Sep 17 00:00:00 2001
From: Ladehunter <ladehunter@gmail.com>
Date: Mon, 18 Sep 2017 03:07:43 +0000
Subject: [PATCH 14/25] Fix bug of ForceOrientation

Signed-off-by: Ladehunter <ladehunter@gmail.com>

Conflicts:
	services/core/java/com/android/server/wm/DisplayContent.java

Change-Id: I3cff4839def38eff08d6e6988c5b99c4bbc4b706
---
 .../com/android/server/wm/DisplayContent.java | 28 +++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/services/core/java/com/android/server/wm/DisplayContent.java b/services/core/java/com/android/server/wm/DisplayContent.java
index 55b3e75c61e..05d34c629a6 100644
--- a/services/core/java/com/android/server/wm/DisplayContent.java
+++ b/services/core/java/com/android/server/wm/DisplayContent.java
@@ -1782,6 +1782,34 @@ class DisplayContent extends WindowContainer<DisplayContent.DisplayChildWindowCo
         mInitialDisplayCutout = mDisplayInfo.displayCutout;
     }
 
+    void getLogicalDisplayRect(Rect out) {
+        // Uses same calculation as in LogicalDisplay#configureDisplayInTransactionLocked.
+        final int orientation = mDisplayInfo.rotation;
+        boolean rotated = (orientation == ROTATION_90 || orientation == ROTATION_270);
+        boolean forcedOrientation = mService.mPolicy.isDefaultOrientationForced();
+        final int physWidth = rotated ? mBaseDisplayHeight : mBaseDisplayWidth;
+        final int physHeight = rotated ? mBaseDisplayWidth : mBaseDisplayHeight;
+        int width = mDisplayInfo.logicalWidth;
+        int left = forcedOrientation ? 0 : (physWidth - width) / 2;
+        int height = mDisplayInfo.logicalHeight;
+        int top = forcedOrientation ? 0 : (physHeight - height) / 2;
+        out.set(left, top, left + width, top + height);
+    }
+
+    private void getLogicalDisplayRect(Rect out, int orientation) {
+        getLogicalDisplayRect(out);
+
+        // Rotate the Rect if needed.
+        final int currentRotation = mDisplayInfo.rotation;
+        final int rotationDelta = deltaRotation(currentRotation, orientation);
+        if (rotationDelta == ROTATION_90 || rotationDelta == ROTATION_270) {
+            createRotationMatrix(rotationDelta, mBaseDisplayWidth, mBaseDisplayHeight, mTmpMatrix);
+            mTmpRectF.set(out);
+            mTmpMatrix.mapRect(mTmpRectF);
+            mTmpRectF.round(out);
+        }
+    }
+
     /**
      * If display metrics changed, overrides are not set and it's not just a rotation - update base
      * values.
-- 
2.17.1

